lib/spack/spack/fetch_strategy.py

```diff
@@ -940,6 +940,125 @@ def __str__(self):
         return '[git] {0}'.format(self._repo_info())
 
 
+@fetcher
+class CvsFetchStrategy(VCSFetchStrategy):
+    """Fetch strategy that gets source code from a CVS repository.
+       Use like this in a package:
+
+           version('name',
+                   cvs=':pserver:anonymous@www.example.com:/cvsroot%module=modulename')
+
+       Optionally, you can provide a branch and/or a date for the URL:
+
+           version('name',
+                   cvs=':pserver:anonymous@www.example.com:/cvsroot%module=modulename',
+                   branch='branchname', date='date')
+
+    Repositories are checked out into the standard stage source path directory.
+    """
+    url_attr = 'cvs'
+    optional_attrs = ['branch', 'date']
+
+    def __init__(self, **kwargs):
+        # Discards the keywords in kwargs that may conflict with the next call
+        # to __init__
+        forwarded_args = copy.copy(kwargs)
+        forwarded_args.pop('name', None)
+        super(CvsFetchStrategy, self).__init__(**forwarded_args)
+
+        self._cvs = None
+        self._expect = None
+        if self.branch is not None:
+            self.branch = str(self.branch)
+        if self.date is not None:
+            self.date = str(self.date)
+
+    @property
+    def cvs(self):
+        if not self._cvs:
+            self._cvs = which('cvs', required=True)
+        return self._cvs
+
+    @property
+    def expect(self):
+        if not self._expect:
+            self._expect = which('expect', required=True)
+        return self._expect
+
+    @property
+    def cachable(self):
+        return self.cache_enabled and (bool(self.branch) or bool(self.date))
+
+    def source_id(self):
+        id = 'id'
+        if self.branch:
+            id += '-branch=' + self.branch
+        if self.date:
+            id += '-date=' + self.date
+        return id
+
+    def mirror_id(self):
+        repo_path = url_util.parse(self.url).path
+        result = os.path.sep.join(['cvs', repo_path])
+        if self.branch:
+            result += '%branch=' + self.branch
+        if self.date:
+            result += '%date=' + self.date
+        return result
+
+    @_needs_stage
+    def fetch(self):
+        if self.stage.expanded:
+            tty.debug('Already fetched {0}'.format(self.stage.source_path))
+            return
+
+        tty.debug('Checking out CVS repository: {0}'.format(self.url))
+
+        with temp_cwd():
+            url, module = self.url.split('%module=')
+            if url.startswith(':pserver:'):
+                # Log in
+                cmds = ['-c', 'spawn cvs -q -d ' + url + ' login',
+                        '-c', 'expect',
+                        '-c', 'send "anonymous\n"',
+                        '-c', 'expect eof']
```

Comment:
(major)

This usage of `expect` is unexpected - according to section 2.9.3.2 of https://www.gnu.org/software/trans-coord/manual/cvs/cvs.html#Remote-repositories you should not have to log in with a password-less user.

I'd like to avoid usage of `expect` if at all possible: for example I'm not clear how these operations affect the CVS client state for subsequent commands (should we be logging out?).

Reply:
There is a file `$HOME/.cvspass` that contains all the ("encrypted") passwords after a login. `cvs logout` would remove the respective password. So yes, this would change global state.

If I recall correctly, in many cases (when using `pserver`) one still has to log in before accessing a remote repository. It's just that the username/password combination is publicly posted on the download web site. See e.g. <https://lorene.obspm.fr/download.html>. At least for this repository, logging in is definitively necessary.

I see in the documentation that one can instead put the password into the download url. I'll try this instead.